src/Lib.P2P/Routing/DistributedQuery.cs

```diff
@@ -0,0 +1,327 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+using MultiFormats;
+using ProtoBuf;
+
+namespace Lib.P2P.Routing
+{
+    /// <summary>
+    ///   A query that is sent to multiple peers.
+    /// </summary>
+    /// <typeparam name="T">
+    ///  The type of answer returned by a peer.
+    /// </typeparam>
+    public sealed class DistributedQuery<T> where T : class
+    {
+        private static readonly ILog Log = LogManager.GetLogger("Lib.P2P.Routing.DistributedQuery");
+        private static int _nextQueryId = 1;
+
+        /// <summary>
+        ///   The maximum number of peers that can be queried at one time
+        ///   for all distributed queries.
+        /// </summary>
+        private static readonly SemaphoreSlim AskCount = new SemaphoreSlim(128);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [A static field in a generic type is not shared among instances of different close constructed types.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)